src/libsyntax/tokenstream.rs

```diff
@@ -186,21 +186,43 @@ impl TokenStream {
     /// Given a `TokenStream` with a `Stream` of only two arguments, return a new `TokenStream`
     /// separating the two arguments with a comma for diagnostic suggestions.
     pub(crate) fn add_comma(&self) -> Option<(TokenStream, Span)> {
-        // Used to suggest if a user writes `println!("{}" a);`
+        // Used to suggest if a user writes `foo!(a b);`
         if let TokenStreamKind::Stream(ref slice) = self.kind {
-            if slice.len() == 2 {
-                let comma_span = match slice[0] {
-                    TokenStream { kind: TokenStreamKind::Tree(TokenTree::Token(sp, _)) } |
-                    TokenStream { kind: TokenStreamKind::Tree(TokenTree::Delimited(sp, _)) } => {
-                        sp.shrink_to_hi()
+            let mut suggestion = None;
+            let mut iter = slice.iter().enumerate().peekable();
+            while let Some((pos, ts)) = iter.next() {
+                if let Some((_, next)) = iter.peek() {
```

Comment:
Neat trick. This solves most of the issues I guess. If it keeps coming up we can consider a generalized version that doesn't hardcode specific tokens

Reply:
For the record, my original idea was to try to do this in https://github.com/rust-lang/rust/blob/3f4f18f098183bf60820e4304433c4c4d9cceaaa/src/libsyntax/ext/tt/macro_rules.rs#L109 and collect all failures with their possible expectations. But I'm not sure what kind of parser changes that would have required.

Reply:
Let's create a follow up ticket. The current approach was slightly less error prone and easier to implement, but that would be more helpful for more macros.